### PR TITLE
PSG-4792: consolidate conda envs

### PR DESCRIPTION
**Pull request by cramshaw**
This PR reduces the docker images the pipeline requires to:
- 1 for the main pipeline and majority of stages
- 1 for pangolin, as it is pinned to python 3.8 and won't play nicely
- 2 for ncov, one illumina, one ONT

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -27,9 +27,12 @@ PyYAML = "5.3.1"
 [tool.poetry.dev-dependencies]
 pytest = "7.1.3"
 
+[tool.poetry.group.dev.dependencies]
+mypy = "^1.8.0"
```
Comment (PatrickJordanCongenica) [inline pyproject.toml:31]: Is it worth moving `pytest` into this group?

File: Makefile [hunk 2]
```diff
@@ -1,6 +1,7 @@
 DOCKER_IMAGE_URI_PATH=566277102435.dkr.ecr.eu-west-2.amazonaws.com/congenica/psga-nonprod
 DOCKER_IMAGE_TAG=dev_latest
 SARS_COV_2=sars_cov_2
+DOCKER_IMAGE_NAME=sars-cov-2-pipeline
```
Comment (PatrickJordanCongenica) [inline Makefile:4]: Is it worth ensuring the name of the Docker image follows the repo name? If so, this should be `pipeline-sars-cov-2`.

File: docker/Dockerfile.psga-pipeline [hunk 3]
```diff
@@ -1,4 +1,4 @@
-FROM condaforge/mambaforge:4.14.0-0 AS condabuild
+FROM condaforge/mambaforge:23.3.1-1 AS condabuild
```
Comment (PatrickJordanCongenica) [inline docker/Dockerfile.psga-pipeline:1]: Do we need to pin to this exact version? Can we relax this to just the minor version so we get security updates for free?

File: docker/Dockerfile.psga-pipeline [hunk 4]
```diff
@@ -7,27 +7,17 @@ RUN mamba env create -f /environment.yml -n custom_env \
...
 
 COPY --from=condabuild /opt/conda/envs/custom_env /opt/conda/envs/custom_env
 ENV PATH=/opt/conda/envs/custom_env/bin:$PATH
 
-# pin the versions of pip and poetry
-ENV PIP_VERSION=22.2.2 \
-    POETRY_VERSION=1.1.13
+# # pin the versions of pip and poetry
+# ENV PIP_VERSION=22.2 \
+#   POETRY_VERSION=1.7
```
Comment (PatrickJordanCongenica) [inline docker/Dockerfile.psga-pipeline:17]: Can we remove this commented out code?

File: docker/Dockerfile.psga-pipeline [hunk 5]
```diff
@@ -42,14 +32,42 @@ WORKDIR /app/psga
 #    This error does not disappear if this is installed via conda
 RUN nextflow plugin install nf-amazon@1.11.0 \
   && pip install setuptools==63.1.0 \
-  && pip install --progress-bar=off --no-cache-dir --upgrade pip==${PIP_VERSION} \
-  && pip install --progress-bar=off --no-cache-dir poetry==${POETRY_VERSION} \
+  && pip install --progress-bar=off --no-cache-dir --upgrade pip \
+  && pip install --progress-bar=off --no-cache-dir poetry \
```
Comment (PatrickJordanCongenica) [inline docker/Dockerfile.psga-pipeline:36]: Just to check, will this just install the latest version of `poetry` without telling it what version to install? If so, won't you end up in a situation where it installs a version of `poetry` that the `poetry.lock` file is incompatible with?

File: docker/Dockerfile.psga-pipeline [hunk 6]
```diff
@@ -42,14 +32,42 @@ WORKDIR /app/psga
...
+
+# scripts for running the validation following integration tests execution
+COPY ./integration_tests/*.py ./app/integration_tests/
+COPY ./integration_tests/expected_data/${pathogen} ./app/integration_tests/expected_data/${pathogen}
+
+# FASTQC
+COPY data/${pathogen}/fastqc/* /
+
+# PRIMER AUTODETECTION
+# COPY scripts/common /app/scripts/common
```
Comment (PatrickJordanCongenica) [inline docker/Dockerfile.psga-pipeline:61]: Commented out?

File: docker/Dockerfile.psga-pipeline [hunk 7]
```diff
@@ -42,14 +32,42 @@ WORKDIR /app/psga
...
+COPY data/${pathogen}/primer_schemes /primer_schemes
+
+# READ IT AND KEEP
+# COPY scripts/common /app/scripts/common
```
Comment (PatrickJordanCongenica) [inline docker/Dockerfile.psga-pipeline:65]: Commented out?

File: docker/sars_cov_2/psga.yml [hunk 8]
```diff
@@ -4,8 +4,26 @@ channels:
...
+  - bcftools=1.16
+  # Read it and Keep
+  - read-it-and-keep=0.3.0
+  # Primer autodetection
+  - biopython=1.80
+  # Pangolin
+  # - pangolin=4.3 # install with pip/git
+  # - pangolin-data=1.21
+  # - scorpio=0.3.17
+  # - constellations=0.1.12
```
Comment (PatrickJordanCongenica) [inline docker/sars_cov_2/psga.yml:29]: Commented out? Is this related to the work to introduce external S3 reference data to this pipeline?

**Review by PatrickJordanCongenica — APPROVED** [review]
LGTM Chris, just a handful of minor comments. A step in the right direction 🦾

File: pyproject.toml (hunk 1 above)
Comment (cramshaw) [reply pyproject.toml:31]: Yes - I want to sort the deps out and get linting properly integrated, but trying to keep the PRs managable.

File: docker/Dockerfile.psga-pipeline (hunk 3 above)
Comment (cramshaw) [reply docker/Dockerfile.psga-pipeline:1]: they don't seem to ship an image that gets security updates, I am unsure how docker would handle it (as these are image tags and not 'real' version numbers)

File: docker/Dockerfile.psga-pipeline (hunk 5 above)
Comment (cramshaw) [reply docker/Dockerfile.psga-pipeline:36]: This is true, hence the pinned numbers above. I wanted to find a more robust way of handling it, as there are conflicting version in the CI too, hence removed for now.

File: docker/sars_cov_2/psga.yml (hunk 8 above)
Comment (cramshaw) [reply docker/sars_cov_2/psga.yml:29]: For reference, these are for pangolin which has had to remain as it's own docker image due to the py3.8 requirement.

**Pull request merged**